crates/matrix-sdk-crypto/src/session_manager/group_sessions/share_strategy.rs

```diff
@@ -528,6 +616,192 @@ mod tests {
+        .await;
+
+        let keys_query = KeyDistributionTestData::dan_keys_query_response();
+        let txn_id = TransactionId::new();
+        machine.mark_request_as_sent(&txn_id, &keys_query).await.unwrap();
+
+        let fake_room_id = room_id!("!roomid:localhost");
+
+        let strategy = CollectStrategy::new_identity_based();
+
+        let encryption_settings =
+            EncryptionSettings { sharing_strategy: strategy.clone(), ..Default::default() };
+
+        let request_result = machine
+            .share_room_key(
+                fake_room_id,
+                vec![KeyDistributionTestData::dan_id()].into_iter(),
+                encryption_settings.clone(),
+            )
+            .await;
+
+        assert!(request_result.is_err());
+        let err = request_result.unwrap_err();
+        match err {
+            OlmError::RoomKeySharingStrategyError(
+                crate::error::RoomKeySharingStrategyError::CrossSigningNotSetup,
+            ) => {
+                // ok
+            }
+            _ => panic!("Unexpected share error"),
+        }
+
+        // Let's now have cross-signing, but the identity is not trusted
+        let keys_query = KeyDistributionTestData::me_keys_query_response();
+        let txn_id = TransactionId::new();
+        machine.mark_request_as_sent(&txn_id, &keys_query).await.unwrap();
+
+        let request_result = machine
+            .share_room_key(
+                fake_room_id,
+                vec![KeyDistributionTestData::dan_id()].into_iter(),
+                encryption_settings.clone(),
+            )
+            .await;
+
+        assert!(request_result.is_err());
+        let err = request_result.unwrap_err();
+        match err {
+            OlmError::RoomKeySharingStrategyError(
+                crate::error::RoomKeySharingStrategyError::SendingFromUnverifiedDevice,
+            ) => {
+                // ok
+            }
+            _ => panic!("Unexpected share error"),
+        }
+
+        // If we are verified it should then work
+        machine
+            .import_cross_signing_keys(CrossSigningKeyExport {
+                master_key: KeyDistributionTestData::MASTER_KEY_PRIVATE_EXPORT.to_owned().into(),
+                self_signing_key: KeyDistributionTestData::SELF_SIGNING_KEY_PRIVATE_EXPORT
+                    .to_owned()
+                    .into(),
+                user_signing_key: KeyDistributionTestData::USER_SIGNING_KEY_PRIVATE_EXPORT
+                    .to_owned()
+                    .into(),
+            })
+            .await
+            .unwrap();
+
+        let request_result = machine
+            .share_room_key(
+                fake_room_id,
+                vec![KeyDistributionTestData::dan_id()].into_iter(),
+                encryption_settings.clone(),
+            )
+            .await;
+
+        assert!(request_result.is_ok());
+    }
+
+    #[async_test]
+    async fn test_share_identity_strategy_report_pinning_violation() {
```

Comment:
as well as a all the cleanups suggested for the other test, this test could do with some comments to explain what's going on.